rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright 2021 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Zeta and related distributions.
+
+use num_traits::Float;
+use crate::{Distribution, Standard};
+use rand::{Rng, distributions::OpenClosed01};
+use core::fmt;
+
+/// Samples floating-point numbers according to the zeta distribution.
+///
+/// The zeta distribution is a limit of the [`Zipf`] distribution.
+///
+/// # Example
+/// ```
+/// use rand::prelude::*;
+/// use rand_distr::Zeta;
+///
+/// let val: f64 = thread_rng().sample(Zeta::new(1.5).unwrap());
+/// println!("{}", val);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    a_minus_1: F,
+    b: F,
+}
+
+/// Error type returned from `Zeta::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ZetaError {
+    /// `a <= 1` or `nan`.
+    ATooSmall,
+}
+
+impl fmt::Display for ZetaError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            ZetaError::ATooSmall => "a <= 1 or is NaN in Zeta distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for ZetaError {}
+
+impl<F> Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    /// Construct a new `Zeta` distribution with given `a` parameter.
+    #[inline]
+    pub fn new(a: F) -> Result<Zeta<F>, ZetaError> {
+        if !(a > F::one()) {
+            return Err(ZetaError::ATooSmall);
+        }
+        let a_minus_1 = a - F::one();
+        let two = F::one() + F::one();
+        Ok(Zeta {
+            a_minus_1,
+            b: two.powf(a_minus_1),
+        })
+    }
+}
+
+impl<F> Distribution<F> for Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
+        // This is based on the numpy implementation.
+        loop {
+            let u = rng.sample(OpenClosed01);
+            let v = rng.sample(Standard);
+            let x = u.powf(-F::one() / self.a_minus_1).floor();
+
+            if x < F::one() {
+                continue;
+            }
+
+            let t = (F::one() + F::one() / x).powf(self.a_minus_1);
+            if v * x * (t - F::one()) / (self.b - F::one()) <= t / self.b {
+                return x;
```

Comment:
We would have to add the transformation from `f64` to `u64` here. If the conversion fails, then loop. Maybe this replaces the previous condition check of `x<1`(?)

Reply:
As a follow up (sorry for the mess), one might want to check the actual value before returning it, taking into account the "effective range" of the sampling algorithm. I am taking into account the following:

- `f64` can not represent all `u64` but only up to `2^(54)` ([stackoverflow discussion](https://stackoverflow.com/questions/12442685/can-a-ieee-754-real-number-cover-all-integers-within-its-range)) continuously and then other isolated integers. 
- The density of `Zeta` is deacreasing, so high values are less likely than smaller values for all parameters `a`.
- Returning high powers of 2 (only because they can be represented in `f64`) would create a strange histogram for very large sampling histograms.

Having said that, I would be in favor of making the "effective range" just 1, 2, ..., 2^(54), instead of 1, 2, ..., u64::MAX. Either way, this is "the best this algorithm can do", no?

Reply:
See below.

Reply:
Okay! I understand.